Impaler Burrower Strain

Shad0vvs opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->

## About The Pull Request

https://streamable.com/4m3rbr (some things are changed since this video)

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->
A burrower strain that can deal damage while burrowed, causing enemies to have to be creative when attacking, removing its weeds will force it to surface.

### Impaler abilities:
**Burrowed Spikes:**
Description: Sends out fast spikes in a direction that hits tile by tile.
Cost: 50 Plasma
Cooldown: 8 Seconds.
Range: 5 Tiles
Damage: 35
**Can be used while burrowed.**

**Sunken Tail:**
Description: After a windup, strikes the targeted area damaging anyone in the tile.
Cost: 100 Plasma
Cooldown: 13 seconds.
Range: 7
Damage: 60
Windup: .8 Seconds

**Ensconce:**
Description: After a windup, go halfway into the ground in a fortified stance (prevents you from moving) which increases your vision range by 3, increases your Burrowed spikes ability range by 2, increases your Sunken Tail ability range by 2, and damage on both abilities by 30.
Modifies Sunken Tail to do an AoE effect:
https://streamable.com/xy0tq1 (middle is full damage + ensconce'ed damage bonus, edges are normal damage)

Cost: None
Cooldown: 3 Seconds
Buffs: Armor deflection increased by 25 and armor explosive resistance increased by 50 and small explosives cannot stun the target while in the stance.

**Base Tail Stab:**
**Can be used while burrowed.**

## Why It's Good For The Game

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding, and may discourage maintainers from reviewing or merging your PR. This section is not strictly required for (non-controversial) fix PRs or backend PRs. -->
Xeno content to tric-I mean to help people have fun when playing Xenos.

## Changelog

<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl: Shad0vvs, Kugamo
add: Added new strain for the Burrower, "Impaler".
fix: Fixed burrowed mobs being able to be targeted by sentries, mines and SG autofire.
fix: Fixed burrowed mobs being able to grab mobs on the surface.
fix: Fixed burrowed mobs being able to resist while burrowed.
fix: Fixed burrowers taking damage from direct flame and shrapnel from explosions.
fix: Fixed burrowers being able to get slashed from enemy Xenos on the surface.
fix: Fixed burrowers unburrow stun to now properly target and stun enemy Xenos.
soundadd: Added sounds for base and Impaler burrower for burrowing and unburrowing, added sounds for Impaler burrower's attacks.
imageadd: Added sprites and effects for Impaler burrower by Kugamo.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->

Shad0vvs commented:
Reminder: Need xenoid sprites before merge.

fira commented on code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:
```diff
@@ -24,6 +24,41 @@
+		mouse_opacity = FALSE
+		playsound(src.loc, 'sound/effects/burrowing_s.ogg', 25)
+		update_icons()
+		addtimer(CALLBACK(src, .proc/do_burrow_cooldown), (caste ? caste.burrow_cooldown : 5 SECONDS))
```
wouldn't they always have a caste ?

Shad0vvs replied:
I blame the burrower coder. will WIP.

fira commented on code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:
```diff
@@ -209,3 +271,270 @@
+	if(!check_plasma_owner())
+		return
+
+	var/turf/target = locate(A.x, A.y, A.z)
```
get_turf

fira commented on code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:
```diff
@@ -209,3 +271,270 @@
+		to_chat(xeno, SPAN_XENOWARNING("You need to do this on weeds!"))
+		return
+
+	if(!check_plasma_owner())
```
isn't plasma cost zero?

if you want to support changing it in the future, might want to also call check_and_use_plasma_owner

Shad0vvs replied:
I want to change plasma after the do_after, just incase someone wants to put a plasma cost on it via ability.
Done via line 460

Shad0vvs commented on code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:
```diff
@@ -24,6 +24,41 @@
 
 	used_burrow = TRUE
 
+	if(mutation_type == BURROWER_IMPALER) //Impaler Burrower specific
```
will wip

Shad0vvs commented:
Made strain specific burrower have onclick burrow ability, modified Ensconce to onclick.
Added xenoid sprites.

Number changes:

Burrowed spikes:
CD from 10 > 8
Damage from 25 > 30
Ensconce'ed range bonus 1 > 2
Ensconce'ed damage bonus 10 > 30

Sunken tail:
CD from 15 > 13
Damage from 40 > 60
Windup delay from 1 Seconds > .8 seconds
Range from 5 > 7

Ensconce'ed range bonus 1 > 2
Ensconce'ed damage bonus 10 > 30
Added ability modification when Ensconce'ed:
https://streamable.com/xy0tq1 (middle is full damage + ensconce'ed damage bonus, edges are normal damage)

Ensconce:
CD from 5 > 3
Vision bonus from 9 > 10

Thegreatspidernonbinary commented:
This is way too fast and gives nearly zero time to react

Shad0vvs commented:
Burrowed Spikes:
Damage 30 > 35

Burrowed checks will happen every 0.5 seconds rather than every 1 seconds to give less leeway if you walk off weeds.

Explosions will affect the burrower now when burrowed, just not with damage. (The stun will be less than normal as they are underground)

Shad0vvs commented:
Update test: Direct sprite clicks will damage the Xeno, otherwise the bullets will fly past them.

fira commented on code/game/objects/effects/aliens.dm:
```diff
@@ -455,6 +455,18 @@
+	mouse_opacity = FALSE
+	var/mob/living/carbon/Xenomorph/source_xeno = null
+
+/obj/effect/xenomorph/ground_spike/New(loc, source_xeno = null, ttl = 10)
```
use Initialize

Shad0vvs commented:
Removed Ensconce being a bodyblock but removed the ability to stack with it.

morrowwolf commented:
Closing as per community ban.
